flow/hstore/hstore.go

```diff
@@ -0,0 +1,240 @@
+
+	// skip to the backslash
+	p.pos = firstBackslash
+
+	// copy bytes until the end, unescaping backslashes
+	for {
+		nextB, end := p.consume()
+		if end {
+			return "", errEOSInQuoted
+		} else if nextB == '"' {
+			break
+		} else if nextB == '\\' {
+			// escape: skip the backslash and copy the char
+			nextB, end = p.consume()
+			if end {
+				return "", errEOSInQuoted
+			}
+			if !(nextB == '\\' || nextB == '"') {
+				return "", fmt.Errorf("unexpected escape in quoted string: found '%#v'", nextB)
+			}
+			builder.WriteByte(nextB)
+		} else {
+			// normal byte: copy it
+			builder.WriteByte(nextB)
+		}
+	}
+	return builder.String(), nil
+}
+
+// consumePairSeparator consumes the Hstore pair separator ", " or returns an error.
+func (p *hstoreParser) consumePairSeparator() error {
+	return p.consumeExpected2(',', ' ')
+}
+
+// consumeKVSeparator consumes the Hstore key/value separator "=>" or returns an error.
+func (p *hstoreParser) consumeKVSeparator() error {
+	return p.consumeExpected2('=', '>')
+}
+
+// consumeDoubleQuotedOrNull consumes the Hstore key/value separator "=>" or returns an error.
+func (p *hstoreParser) consumeDoubleQuotedOrNull() (text, error) {
+	// peek at the next byte
+	if p.atEnd() {
+		return text{}, errors.New("found end instead of value")
+	}
+	next := p.str[p.pos]
+	if next == 'N' {
+		// must be the exact string NULL: use consumeExpected2 twice
+		err := p.consumeExpected2('N', 'U')
+		if err != nil {
+			return text{}, err
+		}
+		err = p.consumeExpected2('L', 'L')
+		if err != nil {
+			return text{}, err
+		}
+		return text{String: "", Valid: false}, nil
+	} else if next != '"' {
+		return text{}, unexpectedByteErr(next, '"')
+	}
+
+	// skip the double quote
+	p.pos += 1
+	s, err := p.consumeDoubleQuoted()
+	if err != nil {
+		return text{}, err
+	}
+	return text{String: s, Valid: true}, nil
+}
+
+func ParseHstore(s string) (string, error) {
+	p := newHSP(s)
+
+	// This is an over-estimate of the number of key/value pairs.
+	numPairsEstimate := strings.Count(s, ">")
+	// makes one allocation of strings for the entire Hstore, rather than one allocation per value.
+	valueStrings := make([]string, 0, numPairsEstimate)
+	result := make(hstore, numPairsEstimate)
+	first := true
+	for !p.atEnd() {
+		if !first {
+			err := p.consumePairSeparator()
+			if err != nil {
+				return "", err
+			}
+		} else {
+			first = false
+		}
+
+		err := p.consumeExpectedByte('"')
+		if err != nil {
+			return "", err
+		}
+
+		key, err := p.consumeDoubleQuoted()
+		if err != nil {
+			return "", err
+		}
+
+		err = p.consumeKVSeparator()
+		if err != nil {
+			return "", err
+		}
+
+		value, err := p.consumeDoubleQuotedOrNull()
+		if err != nil {
+			return "", err
+		}
+		if value.Valid {
+			valueStrings = append(valueStrings, value.String)
+			result[key] = &valueStrings[len(valueStrings)-1]
+		} else {
+			result[key] = nil
+		}
+	}
+
+	// Convert to JSON string
+	jsonString := "{"
```

Comment:
Better to use `encoding/json` here. While it means you need to unescape string val when parsing, it also means you don't have to worry about "or control character" in json spec for strings

https://www.json.org/json-en.html

I assume control characters are https://en.wikipedia.org/wiki/Unicode_control_characters
See also https://en.wikipedia.org/wiki/Control_character